database/mongodb/mongodb.go

```diff
@@ -184,10 +207,70 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
+func (m *Mongo) ensureLockTable() error {
+	indexes := m.db.Collection(m.config.LockingCollection).Indexes()
+	indexOptions := options.Index().SetUnique(true).SetName("lock_unique_key")
+	indexKeys := bsonx.MDoc{
+		DefaultLockingKey: bsonx.Int32(-1),
+	}
+	_, err := indexes.CreateOne(context.TODO(), mongo.IndexModel{
+		Options: indexOptions,
+		Keys:    indexKeys,
+	})
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the MongoDb type.
+func (m *Mongo) ensureVersionTable() (err error) {
+	if err = m.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := m.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if _, _, err = m.Version(); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (m *Mongo) Lock() error {
+	newLockObj := lockObj{
+		Key: lockKey,
+	}
+	_, err := m.db.Collection(m.config.LockingCollection).InsertOne(context.TODO(), newLockObj)
+	if err != nil {
+		return database.ErrLocked
+	}
 	return nil
+
 }
 
 func (m *Mongo) Unlock() error {
+
+	filter := lockObj{
+		lockKey,
+	}
+	_, err := m.db.Collection(m.config.LockingCollection).DeleteMany(context.TODO(), filter)
```

Comment:
What's in context.TODO(). 

I was going to say, if that doesn't have context, maybe we could put more useful information in the collection as well, in additional fields. When the lock was acquired, what our process id is, what our hostname is. 

I don't know what the other systems are doing. In MySQL, because it uses the GET_LOCK(), function you can figure out who holds it. It might be useful for that here in case you need to manually blow away the lock. But maybe that's what context.TODO() is. I dunno, I don't speak go :)

Reply:
```// TODO returns a non-nil, empty Context. Code should use context.TODO when
// it's unclear which Context to use or it is not yet available (because the
// surrounding function has not yet been extended to accept a Context
// parameter).```

Reply:
Ah so then maybe we want to add debug information there.